### PR TITLE
MGDCTRS-1177: Create Subscription and copy Secrets and ConfigMaps

### PR DESCRIPTION
**Pull request by rinaldodev**

### CODE REVIEW

**Review by lburgazzoli — CHANGES_REQUESTED** [review]
We should disable observability and subscription resource copy in dev/local overlay as it would fail as there's no bundle in dev/local and also it may fail in case the observability resource and related CRDs are not installed.

**Comment by rinaldodev** [comment]
> We should disable observability and subscription resource copy in dev/local overlay as it would fail as there's no bundle in dev/local and also it may fail in case the observability resource and related CRDs are not installed.

It was already disabled in the [local overlay](https://github.com/bf2fc6cc711aee1a0c2a/cos-fleetshard/blob/main/etc/kubernetes/manifests/overlays/local/cos-fleetshard-sync/app-config/application.properties#L19). Added the property in dev as well.

File: etc/kubernetes/manifests/overlays/local/observability/crds/customresourcedefinition-subscriptions.operators.coreos.com.yaml [hunk 1]
```diff
@@ -0,0 +1,2636 @@
+﻿kind: CustomResourceDefinition
```
Comment (lburgazzoli) [inline etc/kubernetes/manifests/overlays/local/observability/crds/customresourcedefinition-subscriptions.operators.coreos.com.yaml:1]: is is correct that this CR is in the local overlay ?

File: etc/kubernetes/manifests/overlays/local/observability/crds/customresourcedefinition-subscriptions.operators.coreos.com.yaml [hunk 2]
```diff
@@ -0,0 +1,2636 @@
...
+  name: subscriptions.operators.coreos.com
+  uid: 2442059f-ce23-4d68-8f6d-1c9a6d1c0b6b
+  resourceVersion: '1947'
+  generation: 1
+  creationTimestamp: '2022-05-30T13:00:00Z'
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.6.2
+    include.release.openshift.io/ibm-cloud-managed: 'true'
+    include.release.openshift.io/self-managed-high-availability: 'true'
+  ownerReferences:
```
Comment (lburgazzoli) [inline etc/kubernetes/manifests/overlays/local/observability/crds/customresourcedefinition-subscriptions.operators.coreos.com.yaml:13]: looks like this is an export of a CR, is it correct ?

File: etc/scripts/deploy_observability_crds.sh [hunk 3]
```diff
@@ -1,3 +1,4 @@
 #!/bin/bash -e
 
-kubectl apply -f etc/kubernetes/manifests/overlays/local/observability/crds/observabilities.observability.redhat.com-v1.yml 
+kubectl apply -f etc/kubernetes/manifests/overlays/local/observability/crds/observabilities.observability.redhat.com-v1.yml
```
Comment (lburgazzoli) [inline etc/scripts/deploy_observability_crds.sh:3]: is it correct that such resources are from the local overlay ?

Comment (lburgazzoli) [reply etc/scripts/deploy_observability_crds.sh:3]: ah ok, maybe we can have an it overlay and we can just do an kubectl apply-k of the overlay to get the it set-up, what do  you think ?

File: etc/kubernetes/manifests/overlays/local/observability/crds/customresourcedefinition-subscriptions.operators.coreos.com.yaml (hunk 2 above)
Comment (rinaldodev) [reply etc/kubernetes/manifests/overlays/local/observability/crds/customresourcedefinition-subscriptions.operators.coreos.com.yaml]: Yes it was exported. Cleaned.

File: cos-fleetshard-sync/src/main/java/org/bf2/cos/fleetshard/sync/client/FleetShardObservabilityClient.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            LOGGER.error("Observability CustomResourceDefinition is not present. Observability will not be enabled.");
+            return;
+        }
+
+        if (observabilityNamespace.isBlank()) {
+            LOGGER.error("Observability namespace is not set!");
+            return;
+        }
+
+        if (config.observability().subscription().enabled()) {
```
Comment (lgarciaaco) [inline cos-fleetshard-sync/src/main/java/org/bf2/cos/fleetshard/sync/client/FleetShardObservabilityClient.java:65]: What's the use case where a user sets `enabled` to `true` and `subscription.enabled` to false?

**Review by lgarciaaco — APPROVED** [review]
With some comments but all in all fine

File: etc/scripts/deploy_observability_crds.sh (hunk 3 above)
Comment (rinaldodev) [reply etc/scripts/deploy_observability_crds.sh:3]: @lburgazzoli doing in this one

File: cos-fleetshard-sync/src/main/java/org/bf2/cos/fleetshard/sync/client/FleetShardObservabilityClient.java (hunk 4 above)
Comment (rinaldodev) [reply cos-fleetshard-sync/src/main/java/org/bf2/cos/fleetshard/sync/client/FleetShardObservabilityClient.java:65]: If/when we are able to tell the observability-operator to install the prometheus operator in a different namespace. Then we can (not mandatory) just disable the Subscription and install as a dependency.

File: cos-fleetshard-sync/src/main/java/org/bf2/cos/fleetshard/sync/client/FleetShardObservabilityClient.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+            createSubscriptionResource();
+        }
+        copyResourcesToObservabilityNamespace();
+        createObservabilityResource();
+    }
+
+    private void createSubscriptionResource() {
+        LOGGER.info("Creating Subscription resource");
```
[the remark this answers is not in the document]
Comment (rinaldodev) [reply cos-fleetshard-sync/src/main/java/org/bf2/cos/fleetshard/sync/client/FleetShardObservabilityClient.java:73]: Currently in dev/local observability is disabled, so it won't have an impact right now. I would also consider creating the namespace in the kustomize overlays. That said I'd leave that for later. WDYT?

**Review by lburgazzoli — APPROVED** [review]

**Pull request merged**